Edits ready for review (#95)
cheng-tan/vowpalwabbit.github.io: _guides/getting_started.md

resource_icon: /svg/resources/guide.svg
---

# Getting Started with Linear Regression and Vowpal Wabbit

> This guide describes how to run Linear Regression (LR) algorithms in Vowpal Wabbit (VW). It features an overview of a simple regression problem using a VW workflow tutorial with examples. Find information about interacting with VW, including structuring input, understanding output, and VW diagnostics. The goal is to empower you to explore, experiment, and predict by creating models in VW.

Author: Test

> This guide describes how to run Linear Regression (LR) algorithms in Vowpal Wabbit (VW). It features an overview of a simple regression problem using a VW workflow tutorial with examples. Find information about interacting with VW, including structuring input, understanding output, and VW diagnostics. The goal is to empower you to explore, experiment, and predict by creating models in VW.

Reviewer: ```suggestion
This guide describes how to perform Linear Regression (LR) with Vowpal Wabbit (VW). It features an overview of a simple regression problem using a VW workflow tutorial with examples. Find information about interacting with VW, including structuring input, understanding output, and VW diagnostics. The goal is to empower you to explore, experiment, and predict with VW.
```

Reviewer: Linear Regression is a form of regression, not an algorithm per-se.
Creation of models is one of the many tasks the reader can do, it's not an end goal on itself.
I think "in VW" sounds worse than "with VW" as the result of such regression outlives the usage VW itself.

## Getting started with Vowpal Wabbit

To install VW—and for more information on building VW from source or using a package manager—see [Getting started](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/) guide.

> >**Note:** The CB tutorial uses [Vowpal Wabbit Python package](https://github.com/VowpalWabbit/vowpal_wabbit/tree/master/python). VW is supported in C++ and C#, and additional binary packages are available for select platforms. See [Getting started](https://cheng-tan.github.io/vowpalwabbit.github.io/) module on the homepage for more information.

Reviewer: ```suggestion
>**Note:** The CB tutorial uses [Vowpal Wabbit Python package](https://github.com/VowpalWabbit/vowpal_wabbit/tree/master/python). VW is available from Python, C++ and C#, and additional binary packages are available for select platforms. See [Getting started](https://cheng-tan.github.io/vowpalwabbit.github.io/) module on the homepage for more information.
```

Reviewer: supposed is ambiguous as it might mean that it's available from X but not supported.
Support is a loaded technical term that is usually better replaced.

We want to mention Python, we want to highlight Python.

```
0 | price:.23 sqft:.25 age:.05 2006
```

- The first number in each line is a label. 
- A `0` label corresponds to no roof-replacement, while a `1` label corresponds to a roof-replacement. 
- The bar `|` separates label related data (what we want to predict) from features (what we always know). 
- The features in the first line are `price`, `sqft`, `age`, and `2006`. Each feature may have an optional `:<numeric_value>` following it or, if the value is missing, an implied value of `1`. 

The label information for the second line is more complex than the first: 

`1 2 'second_house | price:.18 sqft:.15 age:.35 1976`

The `1` is the label indicating that a roof replacement is required. 
The `2` is an optional _importance weight_, which implies that this example counts twice. The importance weight comes up in many settings. 

> A missing importance weight defaults to 1. `'second_house` is the tag, it is used elsewhere to identify the example.

Reviewer: I just realized that the wording of `it is used elsewhere to identify the example` is quite vague. It does show up later in this tutorial so shouldn't we setup the right expectation?

```suggestion
A missing importance weight defaults to 1. `'second_house` is the tag, it will show up later to help you identify the example.
```

best constant's loss = 0.250000
total feature number = 15
```
## Vowpal Wabbit diagnostics

This section provides information on the various types of diagnostic information VW presents.

**Use** `--quiet`  command to turn off diagnostic information in Vowpal Wabbit.

### Hash function bits

> The following code sample shows the number hash function bits:

Reviewer: ```suggestion
The following output sample shows the number of bits from the hash function:
```

Reviewer: This is not a code sample, it's from the VW output

```
Num weight bits = 18
```

> This diagnostic shows that the number of hash function bits is 18 (more than enough for this example). Use `-b bits` to adjust the number of bits for a hash function.

Reviewer: ```suggestion
This diagnostic shows that the number of bits from the hash function is 18 (more than enough for this example). Use `-b bits` to adjust the number of bits to be used from the hash function.
```

Reviewer: We use the bits that comes out of the hash function.

### Learning rate

> The following code sample shows the learning rate:

Reviewer: ```suggestion
The following output sample shows the learning rate:
```

Reviewer: This is from the output sample

```
learning rate = 0.5
```

> The default learning rate is `0.5` with current default update (`--normalized --invariant --adaptive`). 

Reviewer: IDK about dropping the mention the reason why 0.5 is the default learning rate.

> If the data is noisy, you need a larger data-set or multiple passes to predict well. For massive data-sets, the learning rate decays to `0` by default. 

Reviewer: ```suggestion
If the data is noisy, you need a larger data-set or multiple passes to predict well. For massive data-sets, the learning rate decays towards `0` by default. 
```

Reviewer: The Leaning rate decays *towards* zero, it doesn't gets *to* zero.

**Use** `-l rate` to adjust the learning rate up or down.

.**Note:** A high learning rate makes the model converge faster than a low learning rate, but if you adjust the learning rate too high, you risk over-fit and end-up worse on average.

### Learning rate decay

> The following code sample shows the initial time:

Reviewer: ```suggestion
The following output sample shows the initial time:
```

```
initial_t = 0
```

> This diagnostic specifies the initial time or duration for learning rate decay. 

Reviewer: ```suggestion
This diagnostic shows the initial time or duration for learning rate decay. 
```
Diagnostic output only shows what was used.

**Use** `--initial_t time` to adjust the initial time. 

> >**Note:** It is not common to adjust the initial time in Vowpal Wabbit

Reviewer: ```suggestion
>**Note:** It is not common to need to adjust the initial time in Vowpal Wabbit
```

Reviewer: We're talking about options that the user might need to use with VW to achieve their goal.
Saying that it's not common doesn't tell anything about why.

### Learning rate decay power

> The following code sample shows the power of learning rate decay:

Reviewer: ```suggestion
The following output sample shows the power of learning rate decay:
```

Not code

```
power_t = 0.5
```

> This diagnostic specifies the power of learning rate decay. The VW default is `0.5` and works well for most problems. 

Reviewer: ```suggestion
This diagnostic shows the power on learning rate decay. The VW default is `0.5` and works well for most problems. 
```

The change from "on" to "of" changes the meaning of the expression. It power in the mathematical sense and not metaphorical one.

The diagnostic output only shows what was used.
